### ISSUE TITLE
Missing type arguments interpreted as unknown

### ISSUE THREAD

**Issue by tkukushkin** [microsoft/pyright#1832]
Such code with `Mapping` without specifying type arguments produce `Mapping[unknown, unknown]` type:
```python
from collections.abc import Mapping

def foo(x: Mapping): ...

foo({1: '2'})
```
Result:
```
  test.py:3:9 - error: Type of parameter "x" is partially unknown
    Parameter type is "Mapping[Unknown, Unknown]" (reportUnknownParameterType)
```

[PEP-484](https://www.python.org/dev/peps/pep-0484/#generics):
> Additionally, `Any` is a valid value for every type variable. Consider the following:
>
> ```
> def count_truthy(elements: List[Any]) -> int:
>     return sum(1 for elem in elements if elem)
> ```
> This is equivalent to omitting the generic notation and just saying `elements: List`.

Mypy in this example produces `Mapping[Any, Any]`. Isn't this a bug of pyright? Is there a way to disable this behavior?

**Comment by tkukushkin** [comment]
Thanks for your reply! It is problem to me because function with `Mapping` annotation is in the third party library and I can't change it, perhaps the author of the library relied on the paragraph from the PEP that I quoted.

I can generate partial stubs for this library but i don't fully understand how I can keep them up to date on libraries updates. On some breaking change I can miss it and check types with old invalid stubs.

**Comment by erictraut** [comment]
Is the library open-source? Perhaps you can contribute back improvements in the form of a PR? We've written some [guidance](https://github.com/microsoft/pyright/blob/master/docs/typed-libraries.md) for library authors on how to improve the types in their libraries.

**Comment by tkukushkin** [comment]
Library is open-source. I can contribute back improvements for some libraries, but unfortunately I don't have that much time to contribute to all libraries, that developers in our company use.

**Comment by tkukushkin** [comment]
I checked that disabling `reportUnknownParameterType` won't hide any errors that I'd like to see. Passing a list instead of a mapping, for example, won't work with the `reportGeneralTypeIssues` error, and that's fine.

I have another question on a similar topic: am I right in understanding that pyright for py.typed packages doesn't analyze the code of those packages itself in any way? For example, I inherit my class from `aiohttp.abc.AbstractAccessLogger` which has no annotation for the `logger` field in the class body. When I use `self.logger` in the `log` method, I get an error `reportUnknownMemberType`. Mypy in this situation, as I understand it, analyzes the `__init__` method code and determines the `logger`, `log_format` field types automatically. Is there anything I can do about this other than make PR in aiohttp? I think there are a lot of such libraries without explicitly defining field types assigned in the constructor , this can make it difficult to actually use pyright. Is it possible to copy mypy's behavior in this case?

```python
from aiohttp import web
from aiohttp.abc import AbstractAccessLogger, BaseRequest


class AccessLogger(AbstractAccessLogger):

    def log(self, request: BaseRequest, response: web.StreamResponse, time: float) -> None:
        reveal_type(self.logger)   # error: Type of "logger" is unknown (reportUnknownMemberType)
```
Pyright output: `info: Type of "self.logger" is "Unknown"`
Mypy output: `note: Revealed type is 'logging.Logger'`

**Comment by erictraut** [comment]
Your assumption is correct. If a package claims to be "py.typed", then pyright assumes that it provides proper type information and does not attempt to infer types. For the reasoning behind this, refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/typed-libraries.md).

You have two options:
1. Delete the "py.typed" file and enable the the "useLibaryCodeForTypes" option. Pyright will then assume that it needs to analyze the library for type information.
2. Work with the package author to improve the type information for the library (e.g. by submitting PRs).

The command-line version of pyright provides a "--verifytypes" mode that allows library maintainers to discover missing type information and determine the "type completeness" of their library.

**Comment by tkukushkin** [comment]
> 1. Delete the "py.typed" file and enable the the "useLibaryCodeForTypes" option. Pyright will then assume that it needs to analyze the library for type information.

This is very inconvenient. Every developer must delete this files

> Work with the package author to improve the type information for the library (e.g. by submitting PRs).

We use a lot of libraries, it's too complicated for us, would require a lot of additional time that mypy doesn't require. Many authors may not know that public attributes must be explicitly annotated, and I don't see any PEP, that requires it.

This behavior stops me from replacing mypy by pyright and may discourage other developers as well.

**Comment by erictraut** [comment]
The `reportUnknownParameterType` diagnostic check is off by default. If you don't like the way this diagnostic check works, simply leave it disabled.

Mypy doesn't track the difference between "Any" and "Unknown". If you would like pyright to match mypy's behavior in this case, don't enable this diagnostic check.

**Comment by tkukushkin** [comment]
I think we no longer understand each other. My last messages are not about `reportUnknownParameterType` at all. I've already turned it off. My problem that pyright doesn't infer types of fields assigned in constructor in third-party libraries, authors of this libraries may not know that they need to explicitly annotate this fields. My feeling is that most libraries have this problem. All libraries that I checked have this problem: aiohttp (library from python core developer), marshmallow, starlette, tornado, pydantic, asyncpg (in https://github.com/MagicStack/asyncpg/pull/577), click, aioredis, aiopg, prompt-toolkit...

**Comment by erictraut** [comment]
@tkukushkin, the issue you're discussing now is completely unrelated to the original issue in this bug report. Could you please open another issue and provide a specific example with repro steps?
